Implement Zero-Knowledge range proofs and number proofs

eanker opened the pull request:
This pr is a first version of the zkp range and number proofs.
There still need to be added some improvements, mostly to the readability/understandably.
But this pr has been opened, so that it can be incorporated with the rest of the code base.

Improvements:
- [ ] Beter Jdoc
- [ ] More tests
- [ ] Improve naming of dataclasses
- [ ] Some code parts can be improved

Made by @eanker and @Recognition2, with a lot of peer programming.

Commit:
Create new ZKP module

Commit:
Add Generator start

Commit:
Borrow ING logic, implement own algorithm

Sorry Liam....
// TODO ING credit geven

eanker committed:
removed example package, extended code and made a test

Commit:
Add NumberProof implementation, include tests. Does work,  not sure how to use as of yet.

The entire exchange between prover and verifier goes via the TrustedParty singleton in this implementation, but this can be changed easily.

Tests for Verifier are added and confirmed working.
Tests for Prover are absent, hard to figure out.
Test coverage still shows 0, not sure how to fix.

Commit:
Eventjes beter loggen, en kood kwalitie omhoog knekkeren

Commit:
Add caching to the prime factorizations

Commit:
Stub for same committed value

Commit:
Rework the ENTIRE Range Proof, implement from scratch.

I have to sleep now.
But one thing that's broken is we need more 'mod' operations everywhere. Some things should be modN but arent

Commit:
Add some print statmenents

Commit:
Add MOD statements where necessary

eanker committed:
Added method for calculating the inverse

Commit:
ZKP is working, needs formal verification but main thing works.

eanker committed:
Asserts now can actually fail

Commit:
Fixed proof that 2 commitments hide the same secret

eanker committed:
Verify that two committments hold the same secret

Commit:
Mid-refactor

Commit:
Nearly fixed, make API a little better yet

Commit:
Remove % operators, fix all tests except SQR tests.

Commit:
Now the proveCommitedNumberIsSquare does not function
Something goes wrong when translating between the implementations of the two papers.

The problem is that on RangeProofProver.kt:74, we SHOULD pass w*w to make cDPrime equal to the commitment created by those parameters, this can be seen better in proveCommitedNumberIsSquare:165.
E = g^w h^r1 modN
But we're passing values such that
E = g^(w*w) h^r1 modN

So that's doomed to fail.
But the thing is, the proveCommittedNumberIsSquare should create the intermediate step, and thus DOES need access to w. It should create an F = g*2 h^r_2
such that E == F^x h^r3.

But I don't know how to fix that, so it's still broken.

The value of r3 is created as r1 - r2*w
I'm not sure why, but because we use it when doing F^x h^r3, it's probably relevant.

Then we're passing all these values  to the proveTwoCommittedIntegersAreEqual, which fails because of an innate property: In that function, E != y1 (one is calculated in the function as g1^x h1^r1, the other one is given as an argument)
But that's correct because it's actually g1^(x^2), because of the square thingy.

I don't even know anymore.

eanker committed:
Fixed a small mistake in the prove is square

eanker committed:
Removed some prints, and verify 100 times

eanker committed:
Changed some formatting

Commit:
Clarity fixes

LiamClark commented on zeroknowledgeproof/build.gradle:
```diff
@@ -0,0 +1,41 @@
+apply plugin: 'com.android.library'
```
We can remove this and apply the normal kotlin plugin here right? or does it have android dependencies?

Recognition2 replied:
No actual android deps afaik, this was generated by AndroidStudio

LiamClark commented on zeroknowledgeproof/src/androidTest/java/com/zeroknowledgeproof/ExampleInstrumentedTest.java:
```diff
@@ -0,0 +1,26 @@
+ *
+ * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
+ */
+@RunWith(AndroidJUnit4.class)
```
let's get rid of this file

LiamClark commented on zeroknowledgeproof/src/main/AndroidManifest.xml:
```diff
@@ -0,0 +1,2 @@
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
```
if we remove hte andorid plugin we can chuck this out too

LiamClark commented on zeroknowledgeproof/src/main/java/com/zeroknowledgeproof/numberProof/NumberProofProver.kt:
```diff
@@ -0,0 +1,47 @@
+import java.math.BigInteger.ZERO
+import java.security.SecureRandom
+
+class NumberProofProver(val that: NumberProofTrustedParty, val N: BigInteger, val givenS: Int?= null) {
```
Even though this pr is for the range proof mostly, we can drop the val for givenS.
we only use it to initialize s, so we don't have to store it as a field so: 

givenS: Int? =null instead of val givenS: Int? =null

LiamClark commented on zeroknowledgeproof/src/main/java/com/zeroknowledgeproof/rangeProof/RangeProofProver.kt:
```diff
@@ -0,0 +1,199 @@
+        //Init
+
+        // Generate security parameters
+//        val k1 = BigInteger(2048, rand)
```
why is this still unused? can we do the security tuning later? but then why is k2 used?

Recognition2 replied:
"Security parameters" are parameters that are used throughout the protocol, but stay the same for a transaction, and are not secret. We need some randomness, this gives it.
k1 is generated later on in L96. The Prover needs to generate it for the later transaction, it's just not necessary in the setup phase.

LiamClark commented on zeroknowledgeproof/src/main/java/com/zeroknowledgeproof/rangeProof/RangeProofProver.kt:
```diff
@@ -0,0 +1,199 @@
+        val pubRes = SetupPublicResult(g = g, h = h, c = c, c1 = c1, c2 = c2, cPrime = cPrime,
+                cDPrime = cDPrime, c1Prime = c1Prime, c2Prime = c2Prime, c3Prime = c3Prime,
+                sameCommitment = proofEqualIntegers, cDPrimeIsSquare = cDPrimeIsSquare,
+                m3IsSquare = m3IsSquare, k1 = BigInteger(2048, rand))
```
ah we initialze k1 here anyways, so I believe we can just throw the one higher up away

LiamClark commented on zeroknowledgeproof/src/main/java/com/zeroknowledgeproof/rangeProof/RangeProofProver.kt:
```diff
@@ -0,0 +1,199 @@
+    fun answerUniqueChallenge(privSetup: SetupPrivateResult, challenge: Challenge): InteractivePublicResult {
+        // Take values from relative structs
+        // private values
+        val (m1, m2, m3, r1, r2, r3) = privSetup
```
If we are destructuring every value of privSetup here anyways to bring them all in scope, why not make this a method on SetupPrivateResult

LiamClark commented on zeroknowledgeproof/src/main/java/com/zeroknowledgeproof/rangeProof/RangeProofProver.kt:
```diff
@@ -0,0 +1,199 @@
+        val b = BigInteger(512, SecureRandom())
+
+
+        //TODO: Hash function H should output 2t-bit strings.
```
is still still a todo or is it solved by the sha-512 down below?

Recognition2 replied:
Not solved by sha512, the proof is too large. t = 80, so the used hash function should output 160 bit strings. SHA512 outputs, well, 512 bit strings.

Recognition2 replied:
We can use SHA256. That'll reduce it to 256 bits, which is still more than "necessary" but I'd rather not use SHA1.

LiamClark replied:
ah so the problem is that it's outputting too many bits right now?

LiamClark commented on zeroknowledgeproof/src/main/java/com/zeroknowledgeproof/rangeProof/RangeProofTrustedParty.kt:
```diff
@@ -0,0 +1,64 @@
+    fun runInteractiveProver(m: Int, a: Int, b: Int) : Boolean {
+        val p = RangeProofProver(m = m, a = a, b = b, N = N)
+        val v = RangeProofVerifier(N = N, low = a, up = b)
+        val (setupPublic, setupPrivate) = p.rangeSetup(this) // Step 1-4
```
nice!

LiamClark commented on zeroknowledgeproof/src/main/java/com/zeroknowledgeproof/rangeProof/RangeProofVerifier.kt:
```diff
@@ -0,0 +1,146 @@
+    fun interactiveVerify(setupRes: SetupPublicResult, answer: InteractivePublicResult): Boolean {
+        // Step 7: Verifier verifies ALL of these claims.
+        // First extract all structs
+        val (c, c1, c2, sameCommitment, cPrime, cDPrime, cDPrimeIsSquare, c1Prime, c2Prime, c3Prime,
```
maybe make this a method on SetupPublicResult

Recognition2 replied:
AFAIK it doesn't really make sense to define one of the verification methods as a method on that data struct

LiamClark commented on zeroknowledgeproof/src/main/java/com/zeroknowledgeproof/rangeProof/RangeProofVerifier.kt:
```diff
@@ -0,0 +1,146 @@
+
+        // Check that nothing is zero
+
+        return c != ZERO &&
```
Let's put some of the conditions in the when clause in seperate functions and let's and them all together in a nicer way than this 50 line return statement :)

Recognition2 replied:
Excellent point.

LiamClark commented on zeroknowledgeproof/src/main/java/com/zeroknowledgeproof/rangeProof/rangeHelperMethods.kt:
```diff
@@ -0,0 +1,136 @@
+val TWO : BigInteger = BigInteger.valueOf(2)
+
+data class SetupPrivateResult(
+        var m1: BigInteger = ZERO,
```
does this one need to be a var?

LiamClark commented on zeroknowledgeproof/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,3 @@
+<resources>
```
can also be removed if we drop the android dependency

LiamClark commented on zeroknowledgeproof/src/test/java/com/zeroknowledgeproof/ExampleUnitTest.kt:
```diff
@@ -0,0 +1,17 @@
+ *
+ * @see [Testing documentation](http://d.android.com/tools/testing)
+ */
+class ExampleUnitTest {
```
chuck it out!

LiamClark reviewed (approved):
Nice to see this working guys!
There's a few questions / cleanup in the comments.

But on a higher level of things (please correct me if i'm wrong this is sort of a guess) We want to put the SetupPublicResult in the blockchain together with the SQR and EL Proofs? so that would be commitedIntegerproof and is there a class for the EL proof?

Let's get this merged asap. don't feel obliged to cleanup all things before we merge. 
@GijsWeterings what do you think?

Recognition2 commented:
Yeah, we want SetupPublicResult on the blockchain, it already contains the SQR and EL proofs.
CommittedIntegerProof IS the EL proof, SQR proof is a translation to an equivalent EL proof.
One thing we didn't figure out is how to preserve the private parameters.
At the moment they are passed back to the TrustedParty, but in reality we'll want the Prover to keep them secret while he awaits the interactive prove-part.
Some sort of a mapping from an ID of the transaction to the corresponding SetupPrivateResult on the prover, or something like that.

The NumberProof contains such a structure, perhaps this is what we want later on?

LiamClark commented:
@recognition2 lets discuss this in issue #6

LiamClark committed:
Remove android plugin form zkp.

This renames the source dir from java to kotlin.
Furthermore move all the asserts to junit4.
tests run in gradle but crash through android studio.
deleted all android related files.

LiamClark committed:
Remove all unused default values for range data classes

LiamClark committed:
Extract function to check for non zero in publicsetup

LiamClark committed:
Move answerUniqueChallenge to Setup Private Result

LiamClark committed:
Throw away commented out k1 code

LiamClark commented:
Great work guys!
